ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/encoding/DepositLogDecoder.java

```diff
@@ -53,52 +56,84 @@ public class DepositLogDecoder {
+    final Bytes signature = extractField(data, SIGNATURE_OFFSET, SIGNATURE_LENGTH);
+    final Bytes index = extractField(data, INDEX_OFFSET, INDEX_LENGTH);
 
+    return Bytes.concatenate(pubKey, withdrawalCred, amount, signature, index);
+  }
+
+  private static void validateLogLength(final Bytes data) {
     if (data.size() != DEPOSIT_LOG_LENGTH) {
-      throw new IllegalArgumentException(
-          "Invalid deposit log length. Must be "
-              + DEPOSIT_LOG_LENGTH
-              + " bytes, but is "
-              + data.size()
-              + " bytes");
+      throw new InvalidDepositLogLayoutException(
+          String.format(
+              "Invalid deposit log length. Must be %d bytes, but is %d bytes",
+              DEPOSIT_LOG_LENGTH, data.size()));
     }
+  }
 
-    final Bytes pubKey = data.slice(DATA_START_POSITION + PUB_KEY_OFFSET, PUB_KEY_LENGTH);
-    final Bytes withdrawalCred =
-        data.slice(DATA_START_POSITION + WITHDRAWAL_CRED_OFFSET, WITHDRAWAL_CRED_LENGTH);
-    final Bytes amount = data.slice(DATA_START_POSITION + AMOUNT_OFFSET, AMOUNT_LENGTH);
-    final Bytes signature = data.slice(DATA_START_POSITION + SIGNATURE_OFFSET, SIGNATURE_LENGTH);
-    final Bytes index = data.slice(DATA_START_POSITION + INDEX_OFFSET, INDEX_LENGTH);
+  private static void validateOffsets(final Bytes data) {
+    validateFieldOffset(data, POSITION_PUB_KEY, PUB_KEY_OFFSET, "pubKey");
+    validateFieldOffset(data, POSITION_WITHDRAWAL_CRED, WITHDRAWAL_CRED_OFFSET, "withdrawalCred");
+    validateFieldOffset(data, POSITION_AMOUNT, AMOUNT_OFFSET, "amount");
+    validateFieldOffset(data, POSITION_SIGNATURE, SIGNATURE_OFFSET, "signature");
+    validateFieldOffset(data, POSITION_INDEX, INDEX_OFFSET, "index");
+  }
 
-    return Bytes.concatenate(pubKey, withdrawalCred, amount, signature, index);
+  private static void validateFieldOffset(
+      final Bytes data, final int position, final int expectedOffset, final String fieldName) {
+    BigInteger offset = data.slice(position, LENGTH_FIELD_SIZE).toBigInteger();
+    if (!offset.equals(BigInteger.valueOf(expectedOffset))) {
+      throw new InvalidDepositLogLayoutException(
+          String.format(
+              "Invalid %s offset: expected %d, but got %d", fieldName, expectedOffset, offset));
+    }
+  }
+
+  private static void validateSizes(final Bytes data) {
+    validateFieldSize(data, PUB_KEY_OFFSET, PUB_KEY_LENGTH, "pubKey");
+    validateFieldSize(data, WITHDRAWAL_CRED_OFFSET, WITHDRAWAL_CRED_LENGTH, "withdrawalCred");
+    validateFieldSize(data, AMOUNT_OFFSET, AMOUNT_LENGTH, "amount");
+    validateFieldSize(data, SIGNATURE_OFFSET, SIGNATURE_LENGTH, "signature");
+    validateFieldSize(data, INDEX_OFFSET, INDEX_LENGTH, "index");
+  }
+
+  private static void validateFieldSize(
+      final Bytes data, final int offset, final int expectedSize, final String fieldName) {
+    BigInteger size = data.slice(offset, LENGTH_FIELD_SIZE).toBigInteger();
```

Comment:
Better to use the size method from bytes, i.e. `data.slice(offset, LENGTH_FIELD_SIZE).size()`

Reply:
We are looking at the content here now the number of bytes